### PR TITLE
Intrinsicify SpanHelpers.IndexOfAny(char,char)

### LINKED ISSUE #12094: Use Intrinsics.X86 for SpanHelpers.IndexOfAny(char,char)
Tracking issue

Implementation available https://github.com/dotnet/coreclr/pull/22187, pending https://github.com/dotnet/coreclr/pull/22505

### PR DESCRIPTION
**Pull request by benaadams**
Updated `Intrinsicify SpanHelpers.IndexOfAny(char,char)` https://github.com/dotnet/coreclr/pull/22877

Resolves: https://github.com/dotnet/runtime/issues/12094

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Char.cs [hunk 1]
```diff
@@ -493,202 +493,240 @@ public static unsafe int IndexOf(ref char searchSpace, char value, int length)
         }
 
         [MethodImpl(MethodImplOptions.AggressiveOptimization)]
-        public static unsafe int IndexOfAny(ref char searchSpace, char value0, char value1, int length)
+        public static int IndexOfAny(ref char searchStart, char value0, char value1, int length)
         {
             Debug.Assert(length >= 0);
+            // If vectors are supported, we first align to Vector.
+            // If the searchSpan has not been fixed or pinned the GC can relocate it during the
+            // execution of this method, so the alignment only acts as best endeavour.
```
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Char.cs:501]: 👍 thanks for the comment (cf. https://github.com/dotnet/runtime/issues/11703).

**Comment by danmosemsft** [comment]
Regex uses IndexOfAny(char, char) and IndexOfAny(char, char, char) so would presumably benefit from this.

**Comment by danmosemsft** [comment]
Same request for perf numbers please - thanks!

**Comment by stephentoub** [comment]
> Regex uses IndexOfAny(char, char) and IndexOfAny(char, char, char) so would presumably benefit from this.

Ideally we'd test this specifically with regex redux, which spends something like 30% of the time now inside of IndexOfAny(char, char).

**Comment by benaadams** [comment]
> Ideally we'd test this specifically with regex redux, which ...

@stephentoub is that in the performance repo?

**Comment by danmosemsft** [comment]
https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/regexredux-csharpcore-5.html

https://github.com/dotnet/performance/blob/master/src/benchmarks/micro/runtime/BenchmarksGame/regex-redux-5.cs

**Comment by benaadams** [comment]
Dealing with the alignment loses most of the benefit, so have updated it to match the approach I took for `SequenceCompareTo` where if it goes vector it does vector for everything (tried a bunch of variations, aligning with loop, unaligned vector => aligned vectors etc), however this worked best

+10%
```
| Faster                                                  | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| ------------------------------------------------------- | ---------:| ----------------:| ----------------:| ------- |
| BenchmarksGame.RegexRedux_1.RunBench                    |      1.11 |      45419250.00 |      41004791.67 |         |
| BenchmarksGame.RegexRedux_5.RunBench(options: Compiled) |      1.10 |       8419890.74 |       7639460.29 | bimodal |
| BenchmarksGame.RegexRedux_5.RunBench(options: None)     |      1.02 |      25805255.56 |      25258357.14 |         |
```

**Comment by danmosemsft** [comment]
10% is significant on that one. This might be worth getting creative to try to get into this release.

@benaadams was that with IndexOfAny(char, char, char) changes as well? I would expect that to be relevant also: https://github.com/dotnet/performance/blob/master/src/benchmarks/micro/runtime/BenchmarksGame/regex-redux-5.cs#L60-L68

**Comment by benaadams** [comment]
Just noticed that's with deleting ` IndexOfAny(char, char, char)` to it uses the very slow path... 😅

**Comment by benaadams** [comment]
Let me put them all in one as they are basically identical methods

**Comment by danmosemsft** [comment]
All your IndexOf PR's into one? That sounds like a good idea, thanks.

**Comment by danmosemsft** [comment]
Your perf numbers above demonstrate that the interpreter doesn't use IndexOfAny at all. @stephentoub had this listed in https://github.com/dotnet/runtime/issues/1349 and it's checked off so maybe he tried it and it didn't give a win somehow.

**Comment by stephentoub** [comment]
I probably checked it off in error, or having employed IndexOf but not IndexOfAny.  There's a TODO in the code for the IndexOfAny (and some other related optimizations):
https://github.com/dotnet/runtime/blob/19a4297925e0ac788b0f3fbebdb13a76241d9281/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexInterpreter.cs#L444-L450

**Comment by danmosemsft** [comment]
OK, I un-checked it for some future day. Presumably the value could only be higher in the interpreter, since the runtime is larger.

**Comment by stephentoub** [comment]
> Presumably the value could only be higher in the interpreter, since the runtime is larger.

Not necessarily.  With the compiler we get to compute the right thing to do once, and there's then no overhead at execution time, since the choice to use IndexOfAny and the arguments to it are all pre-computed.  With the interpreter, we either need to redetermine that every time we invoke FindFirstChar, or we need to compute it once and store that data away somewhere, then check whether we have it and use it.  It would certainly be beneficial when the things being searched for are few and far between, but if the IndexOfAny is used to search for something that frequently presents itself very quickly, it might be a net loss.  Someone would need to do the work and then see whether it's fruitful on a variety of benchmarks.

**Comment by benaadams** [comment]
Superseded by https://github.com/dotnet/runtime/pull/40729

**Pull request closed without merging**